sycl/test-e2e/WorkGroupMemory/reduction_free_function.cpp

```diff
@@ -0,0 +1,266 @@
+// REQUIRES: aspect-usm_shared_allocations
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+// UNSUPPORTED: cuda
+// UNSUPPORTED-INTENDED: The name mangling for free function kernels currently
+// does not work with PTX.
+
+// Usage of work group memory parameters in free function kernels is not yet
+// implemented.
+// TODO: Remove the following directive once
+// https://github.com/intel/llvm/pull/15861 is merged.
+// XFAIL: *
+// XFAIL-TRACKER: https://github.com/intel/llvm/issues/15927
+
+#include <cassert>
+#include <sycl/detail/core.hpp>
+#include <sycl/ext/intel/math.hpp>
+#include <sycl/ext/oneapi/experimental/work_group_memory.hpp>
+#include <sycl/ext/oneapi/free_function_queries.hpp>
+#include <sycl/group_barrier.hpp>
+#include <sycl/usm.hpp>
+
+using namespace sycl;
+
+// Basic usage reduction test using free function kernels.
+// A global buffer is allocated using USM and it is passed to the kernel on the
+// device. On the device, a work group memory buffer is allocated and each item
+// copies the correspondng element of the global buffer to the corresponding
+// element of the work group memory buffer using its global index. The leader of
+// every work-group, after waiting for every work-item to complete, then sums
+// these values storing the result in another work group memory object. Finally,
+// each work item then verifies that the sum of the work group memory elements
+// equals the sum of the global buffer elements. This is repeated for several
+// data types.
+
+queue q;
+context ctx = q.get_context();
+
+constexpr size_t SIZE = 128;
+constexpr size_t VEC_SIZE = 16;
+
+template <typename T>
+void sum_helper(sycl::ext::oneapi::experimental::work_group_memory<T[]> mem,
+                sycl::ext::oneapi::experimental::work_group_memory<T> ret,
+                size_t WGSIZE) {
+  for (int i = 0; i < WGSIZE; ++i) {
+    ret = ret + mem[i];
+  }
+}
+
+template <typename T>
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
+    (ext::oneapi::experimental::nd_range_kernel<1>))
+void sum(sycl::ext::oneapi::experimental::work_group_memory<T[]> mem, T *buf,
+         sycl::ext::oneapi::experimental::work_group_memory<T> result,
+         T expected, size_t WGSIZE, bool UseHelper) {
+  const auto it = sycl::ext::oneapi::this_work_item::get_nd_item<1>();
+  size_t local_id = it.get_local_id();
+  mem[local_id] = buf[local_id];
+  group_barrier(it.get_group());
+  if (it.get_group().leader()) {
+    result = 0;
+    if (!UseHelper) {
+      for (int i = 0; i < WGSIZE; ++i) {
+        result = result + mem[i];
+      }
+    } else {
+      sum_helper(mem, result, WGSIZE);
+    }
+    assert(result == expected);
+  }
+}
+
+// Explicit instantiations for the relevant data types.
```

Comment:
Just curious, why are these explicit instantiations necessary?

Reply:
For some reason, when passing these kernel names to the free function queries with concrete template types, it would complain that these names do not represent kernels because they had not been instantiated yet

Reply:
That sounds like either a bug or a missing feature with free function kernels.  Can you add the same comment here about removing this code when free function kernels are fully supported?

Alternatively, maybe we should just delay testing of `work_group_memory` with free function kernels until there is better support for free function kernels.  It seems like there are a lot of places where you need to add workarounds for missing features in free function kernels.

Reply:
> That sounds like either a bug or a missing feature with free function kernels. Can you add the same comment here about removing this code when free function kernels are fully supported?
> 
> Alternatively, maybe we should just delay testing of `work_group_memory` with free function kernels until there is better support for free function kernels. It seems like there are a lot of places where you need to add workarounds for missing features in free function kernels.

Sure. We can keep the test as I've put an XFAIL directive to expect failures on all platforms since support is not implemented yet.